qt-models/filtermodels.cpp

```diff
@@ -15,605 +15,150 @@
-		if (index >= 0 && items[index].checked)
-			return !negate;
-		head = head->next;
 	}
-	return negate;
-}
 
-void TagFilterModel::diveAdded(const dive *d)
-{
-	struct tag_entry *head = d->tag_list;
-	if (!head) {
-		increaseCount(QString());
-		return;
-	}
-	while (head) {
-		increaseCount(QString());
-		increaseCount(QString(head->tag->name));
-		head = head->next;
-	}
-}
-
-void TagFilterModel::diveDeleted(const dive *d)
-{
-	struct tag_entry *head = d->tag_list;
-	if (!head) {
-		decreaseCount(QString());
-		return;
-	}
-	while (head) {
-		decreaseCount(QString(head->tag->name));
-		head = head->next;
+	// TODO: Finish this iimplementation.
+	bool hasEquipment(const QStringList& equipment, const struct dive *d)
+	{
+		return true;
 	}
 }
 
-BuddyFilterModel::BuddyFilterModel(QObject *parent) : FilterModelBase(parent)
+MultiFilterSortModel *MultiFilterSortModel::instance()
 {
+	static MultiFilterSortModel self;
+	return &self;
 }
 
-int BuddyFilterModel::countDives(const char *s) const
+MultiFilterSortModel::MultiFilterSortModel(QObject *parent) : QSortFilterProxyModel(parent),
+	divesDisplayed(0),
+	curr_dive_site(NULL),
+	model(DiveTripModel::instance())
 {
-	return count_dives_with_person(s);
+	setFilterKeyColumn(-1); // filter all columns
+	setFilterCaseSensitivity(Qt::CaseInsensitive);
+	setSourceModel(model);
 }
 
-static QStringList getDiveBuddies(const dive *d)
+void MultiFilterSortModel::setLayout(DiveTripModel::Layout layout)
 {
-	QString persons = QString(d->buddy) + "," + QString(d->divemaster);
-	QStringList personsList = persons.split(',', QString::SkipEmptyParts);
-	for (QString &s: personsList)
-		s = s.trimmed();
-	return personsList;
+	DiveTripModel *tripModel = DiveTripModel::instance();
+	tripModel->setLayout(layout);	// Note: setLayout() resets the whole model
 }
 
-bool BuddyFilterModel::doFilter(const dive *d) const
+bool MultiFilterSortModel::showDive(const struct dive *d) const
 {
-	// If there's nothing checked, this should show everything
-	// rowCount() == 0 should never happen, because we have the "no tags" row
-	// let's handle it gracefully anyway.
-	if (!anyChecked || rowCount() == 0)
+	if (!m_filterData.validFilter) {
 		return true;
-
-	QStringList personsList = getDiveBuddies(d);
-	// only show empty buddie dives if the user checked that.
-	if (personsList.isEmpty())
-		return items[rowCount() - 1].checked != negate;
-
-	// have at least one buddy
-	// Ignore last item, since this is the "Show Empty Tags" entry
-	for (int i = 0; i < rowCount() - 1; i++) {
-		if (items[i].checked && personsList.contains(items[i].name, Qt::CaseInsensitive))
-			return !negate;
 	}
-	return negate;
-}
 
-void BuddyFilterModel::diveAdded(const dive *d)
-{
-	QStringList buddies = getDiveBuddies(d);
-	if (buddies.empty()) {
-		increaseCount(QString());
-		return;
+	bool shouldShow = true;
+	if (d->visibility < m_filterData.minVisibility || d->visibility > m_filterData.maxVisibility) {
+		shouldShow = false;
```

Comment:
I really dislike this kind of control flow. Why not just do a `return false` here? Shorter and much easier to reason about. This is idiomatic C, because in C you often have to do some cleanups, but in C++ calling the destructors is left to the compiler.